### ISSUE TITLE
Required Param on Stage causes silent failure on Stage run if Param removed

### ISSUE THREAD

**Issue by sygibson** [digitalrebar/provision#1008]
Reporting from #community by `maz` ... 

i noticed something while changing some custom stages earlier - I had a stage with a required param that didn’t exist (I deleted the param and forgot to delete the required entry from the stage). When I ran the workflow containing the stage, it would basically get stuck on the stage with the missing param, but didn’t seem to log anything and just seemed stuck. Took me an age to track that down… is there any logging I’m missing there, or does it wait silently for a particular reason?

**Comment by sygibson** [comment]
Sounds like we don't do any further validation on Objects outside of initial creation time.  It's hard to follow the spider web if a Param is created, to determine what else relies on that Param as a Required object.   

Not sure what the right solution is here ... if the validation checks aren't too expensive, maybe re-validate the dependencies in Workflow prior to running them?  I suspect this could be a pretty expensive operation... 

In either case, better logging at `warn` level may be in order.

**Comment by galthaus** [comment]
This applies to the Required and Optional Parameters going away.

**Comment by zehicle** [comment]
Missing params during execution can cause template render to break in hard to troubleshoot ways.  There are defensive approaches to this, but it's hard to resolve in advance.

**Comment by sygibson** [comment]
Likely best scenario for now is to more accurately catch and log the error that occurs from this.  Params/Stages/etc can change on the fly making this hard to catch early.
